packages/react-form/src/hooks/list/dynamiclist.ts

```diff
@@ -0,0 +1,180 @@
+    }
+  }, [calculatedList, state.initial, dispatch]);
+
+  const validationConfigs = useMemo(
+    () =>
+      mapObject<NormalizedValidationDictionary<any>>(
+        validates,
+        normalizeValidation,
+      ),
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [validates, ...validationDependencies],
+  );
+
+  function addField() {
+    dispatch(addFieldsAction([factory()]));
+  }
+
+  function removeField(index: number) {
+    dispatch(removeFieldsAction(index));
+  }
+
+  const handlers = useMemo(() => {
+    return state.list.map((item, index) => {
+      return mapObject<FieldDictionary<Item>>(
+        item,
+        <Key extends keyof Item & string>(
+          field: FieldState<Item[Key]>,
+          key: Key,
+        ) => {
+          const target = {index, key};
+
+          function validate(value = field.value) {
+            const validates = validationConfigs[key];
+
+            if (validates == null) {
+              return;
+            }
+
+            const siblings = state.list.filter(listItem => listItem !== item);
+
+            return runValidation(
+              error =>
+                dispatch(
+                  updateErrorAction<Item>({target, error: error || ''}),
+                ),
+              {value, siblings, listItem: item},
+              validates,
+            );
+          }
+
+          return {
+            onChange(value: Item[Key] | ChangeEvent) {
+              const normalizedValue = (isChangeEvent(value)
+                ? value.target.value
+                : value) as Item[Key];
+
+              dispatch(
+                updateAction({
+                  target,
+                  value: normalizedValue,
+                }),
+              );
+
+              if (field.error) {
+                validate(normalizedValue);
+              }
+            },
+            reset() {
+              dispatch(resetAction({target}));
+            },
+            newDefaultValue(value: Item[Key]) {
+              dispatch(newDefaultAction({target, value}));
+            },
+            runValidation: validate,
+            onBlur() {
+              const {touched, error} = field;
+
+              if (touched === false && error == null) {
+                return;
+              }
+              validate();
+            },
+            setError(error: string) {
+              dispatch(updateErrorAction({target, error}));
+            },
+          };
+        },
+      );
+    });
+  }, [dispatch, state.list, validationConfigs]);
+
+  const fields: FieldDictionary<Item>[] = useMemo(() => {
+    return state.list.map((item, index) => {
+      return mapObject(item, (field, key: keyof Item) => {
+        return {
+          ...field,
+          ...(handlers[index][key] as any),
+        };
+      });
+    });
+  }, [state.list, handlers]);
+  return {fields, addField, removeField};
+}
+
+function runValidation<Value, Record extends object>(
```

Comment:
Would it be possible to move this function into a utility and share it between the `list` and `dynamiclist`?